identity/src/main/java/com/shotcutter/identity/user/UserIdentityService.java

```diff
@@ -2,39 +2,125 @@
         this.userRepository = userRepository;
         this.converterService = converterService;
+        this.amazonS3 = amazonS3;
     }
 
-    public Optional<UserEntity> findById(String id) {
+    public Mono<UserEntity> findById(String id) {
         return userRepository.findById(id);
     }
 
-    public Optional<UserEntity> findByEmail(String email) {
+    public Mono<UserEntity> findByEmail(String email) {
         return userRepository.findByEmail(email);
     }
 
-    public Optional<UserEntity> registerUser(User unregisteredUser) {
-        return converterService
-                .convertTo(unregisteredUser, UserEntity.class)
-                .map(userRepository::save);
+    public Mono<UserEntity> registerUser(User unregisteredUser) {
+        return Mono.just(converterService.convertTo(unregisteredUser, UserEntity.class))
+                .flatMap(userRepository::save);
+    }
+
+    public Mono<UserEntity> patch(String userId, String newName) {
+        return findById(userId)
+                .map(user -> user.withUsername(newName))
+                .flatMap(userRepository::save);
+    }
+
+    public Mono<UserEntity> updateAvatar(String userId, MultipartFile avatar) {
+        var path = getUserAvatarPath(userId, avatar);
+
+        var userMono = findById(userId);
+
+        // if avatar for this user already exist at s3 - we need to remove it at first
+        userMono.subscribe(this::deleteUserAvatar);
+
+        var putObjReq = PutObjectRequest.builder()
+                .bucket(s3BucketName)
+                .acl(S3CannedAccessControlList.PUBLIC_READ.toString())
+                .key(path)
+                .build();
+
+        CompletableFuture<PutObjectResponse> avatarUpdateResponseMono;
+
+        try {
+            avatarUpdateResponseMono = amazonS3.putObject(putObjReq, AsyncRequestBody.fromBytes(avatar.getBytes()));
+        } catch (IOException e) {
+            return Mono.error(new InvalidUserDataException("Provided avatar file is invalid"));
+        }
+
+        return Mono.zip(userMono, Mono.fromFuture(avatarUpdateResponseMono))
+                .flatMap(tuple -> {
+                    var user = tuple.getT1();
+                    var getUrlRequest = GetUrlRequest.builder()
+                            .bucket(s3BucketName)
+                            .key(path)
+                            .build();
+                    var newAvatarUrl = amazonS3.utilities().getUrl(getUrlRequest).toString();
+
+                    return userRepository.save(user.withAvatar(newAvatarUrl));
+                });
+    }
+
+    private Mono<DeleteObjectResponse> deleteUserAvatar(UserEntity user) {
+        if (user.getAvatar() == null || !user.getAvatar().contains(s3BucketName)) {
+            return Mono.just(null);
+        }
+
+        String avatarPath;
+
+        try {
+            avatarPath = new URL(user.getAvatar()).getPath();
+        } catch (MalformedURLException e) {
+            return Mono.just(null);
+        }
```

Comment:
are you sure you need to return it instead of exception?

Reply:
Honestly it's just a formal error handler, coz there's no case when url wouldn't be parsable.

Reply:
In a worst case we'll get a rubbish string instead of the normal url, but we shouldn't do anything then